File: 0101-0200/0149.Max_Points_on_a_Line.js
//20ms





var maxPoints = function(points) {
  let max = 0;
    // x -> x[0],x[1]  (x1,y1)
    for (const x of points) {
      const slopes = new Map();
    // y -> y[0],y[1]  (x2,y2)
    for (const y of points) {
      if (x === y) continue;
      let slope = Infinity;
      // if (x2 - x1 != 0)
      if (y[0] - x[0] !== 0) {
        // slope = (y2 - y1) / (x2 - x1)
        slope = (y[1] - x[1]) / (y[0] - x[0]);
      }
      if (slopes.has(slope)) {
        slopes.set(slope, slopes.get(slope) + 1);
      } else {
        slopes.set(slope, 1);
      }
      max = Math.max(max, slopes.get(slope));
    }
  }
  return max + 1;
};






//10ms








/**
 * @param {number[][]} points
 * @return {number}
 */
var maxPoints = function(points) {
    // brute force solution
    let n = points.length;
    let maxResult = 1;

    for (let i = 0; i < n; i++) {
        let p1 = points[i];
        let slopeMap = new Map();
        for (let j = i + 1; j < n; j++) {
            let slope;
            let p2 = points[j];

            if(p1[0] === p2[0]) {
                slope = 'inf';
            } else {
                slope = (p2[1] - p1[1]) / (p2[0] - p1[0]);
            }
            let count = slopeMap.get(slope) || 0;
            slopeMap.set(slope, count + 1);
            maxResult = Math.max(maxResult, slopeMap.get(slope) + 1);
        }
    }

    return maxResult;
};

File: 0101-0200/0149.Max_Points_on_a_Line.java
//26ms




class Solution {
    public int maxPoints(int[][] points) {
        int max=0;
        // x -> x[0],x[1]  (x1,y1)
        for(int[] x:points)
        {
            Map<Double,Integer> map= new HashMap<>();
            // y -> y[0],y[1]  (x2,y2)
            for(int[] y:points){
                if(x==y)
                  continue;
                
                double slope=0;
                // if (x2 - x1 == 0) 
                if(y[0]-x[0]==0)
                   slope=Double.POSITIVE_INFINITY; 
                else
                   // slope = (y2 - y1) / (x2 - x1)
                   slope = (y[1]-x[1])/(double)(y[0]-x[0]);
                
                map.put(slope,map.getOrDefault(slope,0)+1);
                max=Math.max(max,map.get(slope));
            }
        }
        return max+1;
    }
}







//6ms








class Solution {
    public int maxPoints(int[][] points) {
        if (points.length <= 2) {
            return points.length;
        }
        int res = 2;
        for (int i = 0; i < points.length; i++) {
            res = Math.max(res, getMax(points, i));
        }
        return res;
    }
    public int getMax(int[][] points, int i) {
        Map<Double, Integer> slope = new HashMap<>();
        int max = 1;
        for (int j = i+1; j < points.length; j++) {
            double s;
            if (points[i][0] == points[j][0]) {
                s = Double.POSITIVE_INFINITY;
            } else if (points[i][1] == points[j][1]) {
                s = 0.0;
            } else {
                s = (double) (points[i][0] - points[j][0]) / (points[i][1] - points[j][1]);
            }
            int count = slope.getOrDefault(s, 1) + 1;
            slope.put(s, count);
            max = Math.max(max, count);
        }
        return max;
    }
}


